tests/contract/test_registry_management.py

```diff
@@ -0,0 +1,325 @@
+    assert notary_pool_length == 1
+
+    # Fast foward
+    mine(web3, 100 - web3.eth.blockNumber)
+
+    # Deregister notary 0
+    smc_handler.deregister_notary(private_key=notary_0_private_key)
+    current_period = web3.eth.blockNumber // smc_handler.config['PERIOD_LENGTH']
+    mine(web3, 1)
+    does_notary_exist = smc_handler.does_notary_exist(notary_0_checksum_address)
+    assert does_notary_exist
+    notary_deregistered_period, notary_pool_index = smc_handler.get_notary_info(
+        notary_0_checksum_address
+    )
+    assert notary_deregistered_period == current_period
+    notary_pool_length = smc_handler.notary_pool_len()
+    assert notary_pool_length == 0
+
+
+def test_deregister_then_register(smc_handler):  # noqa: F811
+    web3 = smc_handler.web3
+    default_gas = smc_handler.config['DEFAULT_GAS']
+
+    (notary_0_private_key, notary_0_canonical_address, notary_0_checksum_address) = \
+        generate_notary_credential(notary_index=0)
+
+    # Register notary 0
+    smc_handler.register_notary(private_key=notary_0_private_key)
+    mine(web3, 1)
+    does_notary_exist = smc_handler.does_notary_exist(notary_0_checksum_address)
+    assert does_notary_exist
+    notary_pool_length = smc_handler.notary_pool_len()
+    assert notary_pool_length == 1
+
+    # Fast foward
+    mine(web3, 100 - web3.eth.blockNumber)
+
+    # Deregister notary 0
+    smc_handler.deregister_notary(private_key=notary_0_private_key)
+    current_period = web3.eth.blockNumber // smc_handler.config['PERIOD_LENGTH']
+    mine(web3, 1)
+    does_notary_exist = smc_handler.does_notary_exist(notary_0_checksum_address)
+    assert does_notary_exist
+    notary_deregistered_period, notary_pool_index = smc_handler.get_notary_info(
+        notary_0_checksum_address
+    )
+    assert notary_deregistered_period == current_period
+    notary_pool_length = smc_handler.notary_pool_len()
+    assert notary_pool_length == 0
+
+    # Register again right away
+    tx_hash = smc_handler.register_notary(private_key=notary_0_private_key)
+    mine(web3, 1)
+    # Check pool remain the same and the transaction consume all gas
+    notary_pool_length = smc_handler.notary_pool_len()
+    assert notary_pool_length == 0
+    assert web3.eth.getTransactionReceipt(tx_hash)['gasUsed'] == default_gas
+
+
+def test_normal_release_notary(smc_handler):  # noqa: F811
+    SIM_NOTARY_LOCKUP_LENGTH = 120
+
+    web3 = smc_handler.web3
+
+    (notary_0_private_key, notary_0_canonical_address, notary_0_checksum_address) = \
+        generate_notary_credential(notary_index=0)
+
+    # Register notary 0
+    smc_handler.register_notary(private_key=notary_0_private_key)
+    mine(web3, 1)
+    does_notary_exist = smc_handler.does_notary_exist(notary_0_checksum_address)
+    assert does_notary_exist
+    notary_pool_length = smc_handler.notary_pool_len()
+    assert notary_pool_length == 1
+
+    # Fast foward
+    mine(web3, 50 - web3.eth.blockNumber)
```

Comment:
Why is this `50 - web3.eth.blockNumber`?

Reply:
Just fast-fowarding to 50-th block. Though there's no particular reason for the number '50'.